File: ifc-tools/src/oebf/ifc_exporter.py
import json
import math
from pathlib import Path

import ifcopenshell
import ifcopenshell.api
import ifcopenshell.api.root
import ifcopenshell.api.unit
import ifcopenshell.api.context
import ifcopenshell.api.project
import ifcopenshell.api.spatial
import ifcopenshell.api.aggregate
import ifcopenshell.api.geometry
import ifcopenshell.api.material
import ifcopenshell.api.pset

# Default wall height when the profile has no explicit height field
WALL_HEIGHT = 2.7


def export_ifc(oebf_dir: Path, ifc_path: Path) -> None:
    manifest = json.loads((oebf_dir / "manifest.json").read_text())
    model_data = json.loads((oebf_dir / "model.json").read_text())
    materials_lib = json.loads((oebf_dir / "materials" / "library.json").read_text())
    mat_by_id = {m["id"]: m for m in materials_lib.get("materials", [])}

    ifc = ifcopenshell.api.project.create_file(version="IFC4")
    project = ifcopenshell.api.root.create_entity(
        ifc, ifc_class="IfcProject", name=manifest["project_name"]
    )
    ifcopenshell.api.unit.assign_unit(ifc)
    body_context = ifcopenshell.api.context.add_context(
        ifc, context_type="Model", context_identifier="Body",
        target_view="MODEL_VIEW",
    )

    site = ifcopenshell.api.root.create_entity(ifc, ifc_class="IfcSite", name="Site")
    building = ifcopenshell.api.root.create_entity(ifc, ifc_class="IfcBuilding", name="Building")
    storey = ifcopenshell.api.root.create_entity(
        ifc, ifc_class="IfcBuildingStorey", name="Ground Floor"
    )
    ifcopenshell.api.aggregate.assign_object(ifc, relating_object=project, products=[site])
    ifcopenshell.api.aggregate.assign_object(ifc, relating_object=site, products=[building])
    ifcopenshell.api.aggregate.assign_object(ifc, relating_object=building, products=[storey])

    for element_id in model_data.get("elements", []):
        _export_element(ifc, oebf_dir, element_id, storey, body_context, mat_by_id)

    for slab_id in model_data.get("slabs", []):
        _export_slab(ifc, oebf_dir, slab_id, storey, body_context, mat_by_id)

    ifc.write(str(ifc_path))


def _export_element(ifc, oebf_dir, element_id, storey, body_context, mat_by_id):
    try:
        elem = json.loads((oebf_dir / "elements" / f"{element_id}.json").read_text())
        path_data = json.loads((oebf_dir / "paths" / f"{elem['path_id']}.json").read_text())
        profile_data = json.loads(
            (oebf_dir / "profiles" / f"{elem['profile_id']}.json").read_text()
        )
    except (FileNotFoundError, KeyError) as exc:
        print(f"  Warning: skipping {element_id}: {exc}")
        return

    entity = ifcopenshell.api.root.create_entity(
        ifc, ifc_class=elem["ifc_type"], name=elem["description"]
    )

    path_length, tangent = _path_length_and_tangent(path_data)
    start = path_data["segments"][0]["start"]
    total_width = sum(layer["thickness"] for layer in profile_data["assembly"])
    origin_x = profile_data.get("origin", {}).get("x", total_width / 2)
    height = WALL_HEIGHT

    _assign_body_geometry(ifc, entity, body_context,
                          start, tangent, path_length, total_width, origin_x, height)
    _assign_material_layers(ifc, entity, profile_data["assembly"], mat_by_id)
    _assign_property_set(ifc, entity, elem.get("properties", {}))

    ifcopenshell.api.spatial.assign_container(ifc, relating_structure=storey, products=[entity])


def _path_length_and_tangent(path_data):
    """Return (length, normalised tangent dict) from the first segment of the path."""
    seg = path_data["segments"][0]
    s, e = seg["start"], seg["end"]
    dx = e["x"] - s["x"]
    dy = e["y"] - s["y"]
    dz = e.get("z", 0.0) - s.get("z", 0.0)
    length = math.sqrt(dx * dx + dy * dy + dz * dz)
    if length < 1e-9:
        return 1.0, {"x": 1.0, "y": 0.0, "z": 0.0}
    return length, {"x": dx / length, "y": dy / length, "z": dz / length}


def _assign_body_geometry(ifc, entity, context,
                          start, tangent, path_length, total_width, origin_x, height):
    """
    Create an IfcExtrudedAreaSolid for the wall body and assign it to the entity.

    The wall footprint (XY plane of the solid's local position) is a rectangle:
      - X axis  = path tangent (wall length direction)
      - Y axis  = wall normal (across thickness)
      - Depth   = wall height (Z direction)
    The rectangle is path_length × total_width, offset in Y so the centreline
    aligns with origin_x.
    """
    tx, ty = tangent["x"], tangent["y"]

    # Wall normal: rotate tangent 90° CCW → (-ty, tx)
    ny, nx_n = tx, -ty  # normal X component, normal Y component
    # Actually: rotate (tx,ty) by 90° CCW → (-ty, tx)
    nnx, nny = -ty, tx

    # ObjectPlacement: origin at path start, X along tangent, Z up
    matrix = [
        [tx,   nnx,  0.0, start["x"]],
        [ty,   nny,  0.0, start["y"]],
        [0.0,  0.0,  1.0, start.get("z", 0.0)],
    ]
    ifcopenshell.api.geometry.edit_object_placement(
        ifc, product=entity, matrix=matrix
    )

    # Profile cross-section in the YZ plane (perpendicular to wall length).
    # The solid's Position uses identity axes relative to ObjectPlacement, so:
    #   profile X → ObjectPlacement Y (across wall thickness)
    #   profile Y → ObjectPlacement Z (height / up)
    # Extrusion is along ObjectPlacement X (wall length direction), Depth = path_length.
    y_offset = -(origin_x - total_width / 2)  # offset from centreline in profile X (across)
    profile = ifc.create_entity(
        "IfcRectangleProfileDef",
        ProfileType="AREA",
        Position=ifc.create_entity(
            "IfcAxis2Placement2D",
            Location=ifc.create_entity(
                "IfcCartesianPoint", Coordinates=[y_offset, height / 2]
            ),
        ),
        XDim=total_width,
        YDim=height,
    )

    solid = ifc.create_entity(
        "IfcExtrudedAreaSolid",
        SweptArea=profile,
        Position=ifc.create_entity(
            "IfcAxis2Placement3D",
            Location=ifc.create_entity("IfcCartesianPoint", Coordinates=[0.0, 0.0, 0.0]),
        ),
        ExtrudedDirection=ifc.create_entity("IfcDirection", DirectionRatios=[1.0, 0.0, 0.0]),
        Depth=path_length,
    )

    shape_rep = ifc.create_entity(
        "IfcShapeRepresentation",
        ContextOfItems=context,
        RepresentationIdentifier="Body",
        RepresentationType="SweptSolid",
        Items=[solid],
    )
    entity.Representation = ifc.create_entity(
        "IfcProductDefinitionShape", Representations=[shape_rep]
    )


def _assign_material_layers(ifc, entity, assembly, mat_by_id):
    """Create an IfcMaterialLayerSet from the profile assembly and assign it."""
    ifc_layers = []
    for layer in assembly:
        mat_info = mat_by_id.get(layer["material_id"], {})
        ifc_mat = ifc.create_entity("IfcMaterial", Name=mat_info.get("name", layer["material_id"]))
        ifc_layer = ifc.create_entity(
            "IfcMaterialLayer",
            Material=ifc_mat,
            LayerThickness=layer["thickness"],
            Name=layer["name"],
        )
        ifc_layers.append(ifc_layer)

    layer_set = ifc.create_entity("IfcMaterialLayerSet", MaterialLayers=ifc_layers)
    layer_set_usage = ifc.create_entity(
        "IfcMaterialLayerSetUsage",
        ForLayerSet=layer_set,
        LayerSetDirection="AXIS2",
        DirectionSense="POSITIVE",
        OffsetFromReferenceLine=0.0,
    )
    ifc.create_entity(
        "IfcRelAssociatesMaterial",
        GlobalId=ifcopenshell.guid.new(),
        RelatedObjects=[entity],
        RelatingMaterial=layer_set_usage,
    )


def _export_slab(ifc, oebf_dir, slab_id, storey, body_context, mat_by_id):
    """Export one OEBF Slab as IfcSlab with IfcArbitraryClosedProfileDef geometry."""
    try:
        slab = json.loads((oebf_dir / "slabs" / f"{slab_id}.json").read_text())
        path_data = json.loads(
            (oebf_dir / "paths" / f"{slab['boundary_path_id']}.json").read_text()
        )
    except (FileNotFoundError, KeyError) as exc:
        print(f"  Warning: skipping slab {slab_id}: {exc}")
        return

    entity = ifcopenshell.api.root.create_entity(
        ifc, ifc_class=slab["ifc_type"], name=slab["description"]
    )

    pts_2d = [
        (seg["start"]["x"], seg["start"]["y"])
        for seg in path_data["segments"]
        if seg["type"] == "line"
    ]

    elevation = slab.get("elevation_m", 0.0)
    thickness = slab["thickness_m"]

    matrix = [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, elevation],
    ]
    ifcopenshell.api.geometry.edit_object_placement(ifc, product=entity, matrix=matrix)

    ifc_pts = [ifc.create_entity("IfcCartesianPoint", Coordinates=(float(p[0]), float(p[1]))) for p in pts_2d]
    ifc_pts.append(ifc_pts[0])  # close the polyline
    polyline = ifc.create_entity("IfcPolyline", Points=ifc_pts)
    profile = ifc.create_entity(
        "IfcArbitraryClosedProfileDef",
        ProfileType="AREA",
        OuterCurve=polyline,
    )

    solid = ifc.create_entity(
        "IfcExtrudedAreaSolid",
        SweptArea=profile,
        Position=ifc.create_entity(
            "IfcAxis2Placement3D",
            Location=ifc.create_entity("IfcCartesianPoint", Coordinates=[0.0, 0.0, 0.0]),
        ),
        ExtrudedDirection=ifc.create_entity("IfcDirection", DirectionRatios=[0.0, 0.0, -1.0]),
        Depth=thickness,
    )

    shape_rep = ifc.create_entity(
        "IfcShapeRepresentation",
        ContextOfItems=body_context,
        RepresentationIdentifier="Body",
        RepresentationType="SweptSolid",
        Items=[solid],
    )
    entity.Representation = ifc.create_entity(
        "IfcProductDefinitionShape", Representations=[shape_rep]
    )

    mat_info = mat_by_id.get(slab.get("material_id", ""), {})
    if mat_info:
        ifc_mat = ifc.create_entity("IfcMaterial", Name=mat_info.get("name", slab["material_id"]))
        ifc.create_entity(
            "IfcRelAssociatesMaterial",
            GlobalId=ifcopenshell.guid.new(),
            RelatedObjects=[entity],
            RelatingMaterial=ifc_mat,
        )

    _assign_property_set(ifc, entity, slab.get("properties", {}))
    ifcopenshell.api.spatial.assign_container(ifc, relating_structure=storey, products=[entity])


def _assign_property_set(ifc, entity, properties: dict):
    """Write a single IfcPropertySet named OEBF_Properties from the element properties dict."""
    if not properties:
        return
    pset = ifcopenshell.api.pset.add_pset(ifc, product=entity, name="OEBF_Properties")
    props = {}
    for key, val in properties.items():
        if isinstance(val, bool):
            props[key] = val
        elif isinstance(val, int):
            props[key] = val
        elif isinstance(val, float):
            props[key] = val
        else:
            props[key] = str(val)
    ifcopenshell.api.pset.edit_pset(ifc, pset=pset, properties=props)

File: viewer/src/ifc/ifcExporter.js
/**
 * ifcExporter.js — Export an OEBF bundle to IFC4 STEP text.
 *
 * Usage:
 *   const ifcText = await exportBundleToIfc(adapter);
 *   // trigger download of ifcText as a .ifc file
 */

const WALL_HEIGHT = 2.7; // default wall height (metres)

/**
 * Export the current bundle to IFC4 STEP format.
 *
 * @param {object} adapter  — FsaAdapter or MemoryAdapter
 * @returns {Promise<string>} IFC STEP file content
 */
export async function exportBundleToIfc(adapter) {
  const manifest = await adapter.readJson('manifest.json');

  let model = { elements: [], slabs: [] };
  try { model = await adapter.readJson('model.json'); } catch { /* empty bundle */ }

  let matsLib = { materials: [] };
  try { matsLib = await adapter.readJson('materials/library.json'); } catch { /* no materials */ }
  const matById = Object.fromEntries(matsLib.materials.map(m => [m.id, m]));

  const gen = new IfcGen();

  // ── Hierarchy ──────────────────────────────────────────────────────────────
  const unitsId   = gen.units();
  const ctxId     = gen.context();
  const projectId = gen.project(manifest.project_name || 'OEBF Project', unitsId, ctxId);
  const siteId    = gen.site('Site');
  const buildId   = gen.building('Building');
  const storeyId  = gen.storey('Ground Floor', 0.0);

  gen.aggregate(projectId, [siteId]);
  gen.aggregate(siteId,    [buildId]);
  gen.aggregate(buildId,   [storeyId]);

  const contained = [];

  // ── Elements ───────────────────────────────────────────────────────────────
  for (const elemId of (model.elements ?? [])) {
    try {
      const elem    = await adapter.readJson(`elements/${elemId}.json`);
      const path    = await adapter.readJson(`paths/${elem.path_id}.json`);
      let   profile = null;
      try { profile = await adapter.readJson(`profiles/${elem.profile_id}.json`); } catch { /* no profile */ }

      const id = gen.wall(elem, path, profile, matById, ctxId);
      if (id) contained.push(id);
    } catch (e) {
      console.warn(`IFC export: skipping element ${elemId}: ${e.message}`);
    }
  }

  // ── Slabs ──────────────────────────────────────────────────────────────────
  for (const slabId of (model.slabs ?? [])) {
    try {
      const slab = await adapter.readJson(`slabs/${slabId}.json`);
      const path = await adapter.readJson(`paths/${slab.boundary_path_id}.json`);

      const id = gen.slab(slab, path, matById, ctxId);
      if (id) contained.push(id);
    } catch (e) {
      console.warn(`IFC export: skipping slab ${slabId}: ${e.message}`);
    }
  }

  if (contained.length > 0) gen.contain(storeyId, contained);

  return gen.toString(manifest.project_name || 'model');
}

// ── IFC entity generator ───────────────────────────────────────────────────

class IfcGen {
  constructor() {
    this._lines  = [];
    this._nextId = 1;
  }

  _id()          { return this._nextId++; }
  _ref(id)       { return `#${id}`; }
  _refs(ids)     { return `(${ids.map(i => '#' + i).join(',')})` || '()'; }

  _emit(id, type, ...attrs) {
    this._lines.push(`#${id}=${type}(${attrs.join(',')});`);
    return id;
  }

  // ── Geometric helpers ────────────────────────────────────────────────────

  pt3(x, y, z) {
    return this._emit(this._id(), 'IFCCARTESIANPOINT', `(${_f(x)},${_f(y)},${_f(z)})`);
  }

  pt2(x, y) {
    return this._emit(this._id(), 'IFCCARTESIANPOINT', `(${_f(x)},${_f(y)})`);
  }

  dir3(x, y, z) {
    return this._emit(this._id(), 'IFCDIRECTION', `(${_f(x)},${_f(y)},${_f(z)})`);
  }

  axis2place3d(locId, axisId = null, refDirId = null) {
    return this._emit(this._id(), 'IFCAXIS2PLACEMENT3D',
      this._ref(locId),
      axisId   ? this._ref(axisId)   : '$',
      refDirId ? this._ref(refDirId) : '$',
    );
  }

  axis2place2d(locId, dirId = null) {
    return this._emit(this._id(), 'IFCAXIS2PLACEMENT2D',
      this._ref(locId),
      dirId ? this._ref(dirId) : '$',
    );
  }

  localPlace(placeId) {
    return this._emit(this._id(), 'IFCLOCALPLACEMENT', '$', this._ref(placeId));
  }

  // ── Top-level entities ───────────────────────────────────────────────────

  units() {
    const si = this._emit(this._id(), 'IFCSIUNIT', '*', '.LENGTHUNIT.', '$', '.METRE.');
    return this._emit(this._id(), 'IFCUNITASSIGNMENT', `(${this._ref(si)})`);
  }

  context() {
    const loc    = this.pt3(0, 0, 0);
    const place  = this.axis2place3d(loc);
    return this._emit(this._id(), 'IFCGEOMETRICREPRESENTATIONCONTEXT',
      '$', "'Model'", '3', '1.E-05', this._ref(place), '$');
  }

  project(name, unitsId, ctxId) {
    return this._emit(this._id(), 'IFCPROJECT',
      `'${_guid()}'`, '$', `'${_esc(name)}'`, '$', '$', '$', '$',
      `(${this._ref(ctxId)})`, this._ref(unitsId));
  }

  site(name) {
    const loc   = this.pt3(0, 0, 0);
    const pl    = this.axis2place3d(loc);
    const lp    = this.localPlace(pl);
    return this._emit(this._id(), 'IFCSITE',
      `'${_guid()}'`, '$', `'${_esc(name)}'`, '$', '$',
      this._ref(lp), '$', '$', '.ELEMENT.', '$', '$', '$', '$');
  }

  building(name) {
    const loc = this.pt3(0, 0, 0);
    const pl  = this.axis2place3d(loc);
    const lp  = this.localPlace(pl);
    return this._emit(this._id(), 'IFCBUILDING',
      `'${_guid()}'`, '$', `'${_esc(name)}'`, '$', '$',
      this._ref(lp), '$', '$', '.ELEMENT.', '$', '$', '$');
  }

  storey(name, elev) {
    const loc = this.pt3(0, 0, elev);
    const pl  = this.axis2place3d(loc);
    const lp  = this.localPlace(pl);
    return this._emit(this._id(), 'IFCBUILDINGSTOREY',
      `'${_guid()}'`, '$', `'${_esc(name)}'`, '$', '$',
      this._ref(lp), '$', '$', '.ELEMENT.', `${_f(elev)}`);
  }

  aggregate(parentId, childIds) {
    return this._emit(this._id(), 'IFCRELAGGREGATES',
      `'${_guid()}'`, '$', '$', '$',
      this._ref(parentId), this._refs(childIds));
  }

  contain(storeyId, productIds) {
    if (!productIds.length) return;
    return this._emit(this._id(), 'IFCRELCONTAINEDINSPATIALSTRUCTURE',
      `'${_guid()}'`, '$', "'Physical model'", '$',
      this._refs(productIds), this._ref(storeyId));
  }

  // ── Element geometry ─────────────────────────────────────────────────────

  wall(elem, pathData, profile, matById, ctxId) {
    const seg = pathData.segments?.[0];
    if (!seg) return null;

    const s = seg.start ?? { x: 0, y: 0, z: 0 };
    const e = seg.end   ?? { x: s.x + 1, y: s.y, z: s.z };

    const dx = e.x - s.x, dy = e.y - s.y, dz = (e.z ?? 0) - (s.z ?? 0);
    const length = Math.sqrt(dx*dx + dy*dy + dz*dz) || 1;
    const tx = dx / length, ty = dy / length, tz = dz / length;

    const totalWidth = profile?.assembly?.reduce((sum, l) => sum + l.thickness, 0) ?? 0.3;
    const wallHeight = profile?.height_limit_m ?? WALL_HEIGHT;
    const originX    = profile?.origin?.x ?? (totalWidth / 2);

    // Object placement: origin at wall start, RefDir along tangent, Axis = Z-up
    const originPt = this.pt3(s.x ?? 0, s.y ?? 0, s.z ?? 0);
    const zAxis    = this.dir3(0, 0, 1);
    const refDir   = this.dir3(tx, ty, tz);
    const place3d  = this.axis2place3d(originPt, zAxis, refDir);
    const lp       = this.localPlace(place3d);

    // Rectangle profile in local YZ plane (perpendicular to wall length)
    const yOff      = -(originX - totalWidth / 2);
    const profOrigin = this.pt2(yOff, wallHeight / 2);
    const profPlace  = this.axis2place2d(profOrigin);
    const rectProf   = this._emit(this._id(), 'IFCRECTANGLEPROFILEDEF',
      '.AREA.', '$', this._ref(profPlace), `${_f(totalWidth)}`, `${_f(wallHeight)}`);

    // Extruded solid along local X (wall length)
    const solidOrigin = this.pt3(0, 0, 0);
    const solidPlace  = this.axis2place3d(solidOrigin);
    const extrudeDir  = this.dir3(1, 0, 0);
    const solid = this._emit(this._id(), 'IFCEXTRUDEDAREASOLID',
      this._ref(rectProf), this._ref(solidPlace), this._ref(extrudeDir), `${_f(length)}`);

    // Shape representation
    const shapeRep = this._emit(this._id(), 'IFCSHAPEREPRESENTATION',
      this._ref(ctxId), "'Body'", "'SweptSolid'", `(${this._ref(solid)})`);
    const prodShape = this._emit(this._id(), 'IFCPRODUCTDEFINITIONSHAPE',
      '$', '$', `(${this._ref(shapeRep)})`);

    // Wall entity
    const stepType = (elem.ifc_type || 'IfcWall').toUpperCase();
    const wallId   = this._emit(this._id(), stepType,
      `'${_guid()}'`, '$', `'${_esc(elem.description || 'Wall')}'`, '$', '$',
      this._ref(lp), this._ref(prodShape), '$', '$.NOTDEFINED.$');

    // Material layer set
    if (profile?.assembly?.length) {
      const layerIds = profile.assembly.map(layer => {
        const matInfo = matById[layer.material_id] ?? {};
        const matId   = this._emit(this._id(), 'IFCMATERIAL',
          `'${_esc(matInfo.name || layer.material_id)}'`, '$', '$');
        return this._emit(this._id(), 'IFCMATERIALLAYER',
          this._ref(matId), `${_f(layer.thickness)}`, '$',
          `'${_esc(layer.name)}'`, '$', '$', '$');
      });
      const layerSet = this._emit(this._id(), 'IFCMATERIALLAYERSET',
        this._refs(layerIds), `'${_esc(elem.description || 'Wall')}'`, '$');
      const usage = this._emit(this._id(), 'IFCMATERIALLAYERSETUSAGE',
        this._ref(layerSet), '.AXIS2.', '.POSITIVE.', '0.');
      this._emit(this._id(), 'IFCRELASSOCIATESMATERIAL',
        `'${_guid()}'`, '$', '$', '$',
        `(${this._ref(wallId)})`, this._ref(usage));
    }

    return wallId;
  }

  slab(slabData, pathData, matById, ctxId) {
    const elevation = slabData.elevation_m ?? 0;
    const thickness = slabData.thickness_m ?? 0.2;

    const pts = pathData.segments
      ?.filter(s => s.type === 'line')
      ?.map(s => s.start) ?? [];
    if (pts.length < 3) return null;

    // Object placement at slab elevation
    const originPt = this.pt3(0, 0, elevation);
    const place3d  = this.axis2place3d(originPt);
    const lp       = this.localPlace(place3d);

    // Boundary polyline (closed)
    const ptIds = pts.map(p => this.pt2(p.x, p.y));
    ptIds.push(ptIds[0]); // close the loop
    const polyline = this._emit(this._id(), 'IFCPOLYLINE', this._refs(ptIds));
    const slabProf = this._emit(this._id(), 'IFCARBITRARYCLOSEDPROFILEDEF',
      '.AREA.', '$', this._ref(polyline));

    // Extrude downward
    const solidOrigin = this.pt3(0, 0, 0);
    const solidPlace  = this.axis2place3d(solidOrigin);
    const extrudeDir  = this.dir3(0, 0, -1);
    const solid = this._emit(this._id(), 'IFCEXTRUDEDAREASOLID',
      this._ref(slabProf), this._ref(solidPlace), this._ref(extrudeDir), `${_f(thickness)}`);

    // Shape representation
    const shapeRep = this._emit(this._id(), 'IFCSHAPEREPRESENTATION',
      this._ref(ctxId), "'Body'", "'SweptSolid'", `(${this._ref(solid)})`);
    const prodShape = this._emit(this._id(), 'IFCPRODUCTDEFINITIONSHAPE',
      '$', '$', `(${this._ref(shapeRep)})`);

    // Slab entity
    const stepType = (slabData.ifc_type || 'IfcSlab').toUpperCase();
    const slabId   = this._emit(this._id(), stepType,
      `'${_guid()}'`, '$', `'${_esc(slabData.description || 'Slab')}'`, '$', '$',
      this._ref(lp), this._ref(prodShape), '$', '$.FLOOR.$');

    // Material
    const matInfo = matById[slabData.material_id] ?? {};
    if (matInfo.name) {
      const matId = this._emit(this._id(), 'IFCMATERIAL',
        `'${_esc(matInfo.name)}'`, '$', '$');
      this._emit(this._id(), 'IFCRELASSOCIATESMATERIAL',
        `'${_guid()}'`, '$', '$', '$',
        `(${this._ref(slabId)})`, this._ref(matId));
    }

    return slabId;
  }

  // ── Output ───────────────────────────────────────────────────────────────

  toString(projectName) {
    const now = new Date().toISOString().replace('T', 'T').slice(0, 19);
    return [
      'ISO-10303-21;',
      'HEADER;',
      `FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');`,
      `FILE_NAME('${_esc(projectName)}.ifc','${now}',(''),(''),'OEBF Editor','OEBF v0.3','');`,
      `FILE_SCHEMA(('IFC4'));`,
      'ENDSEC;',
      'DATA;',
      ...this._lines,
      'ENDSEC;',
      'END-ISO-10303-21;',
    ].join('\n');
  }
}

// ── Utility functions ──────────────────────────────────────────────────────

const IFC_GUID_CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_$';

function _guid() {
  const bytes = new Uint8Array(16);
  (typeof crypto !== 'undefined' ? crypto : { getRandomValues: b => { for (let i = 0; i < b.length; i++) b[i] = Math.floor(Math.random() * 256); } }).getRandomValues(bytes);
  let n = 0n;
  for (const b of bytes) n = (n << 8n) | BigInt(b);
  let result = '';
  for (let i = 0; i < 22; i++) {
    result = IFC_GUID_CHARS[Number(n % 64n)] + result;
    n /= 64n;
  }
  return result;
}

function _f(n) {
  if (n == null || isNaN(n)) return '0.';
  const f = parseFloat(n);
  if (f === 0) return '0.';
  const s = f.toFixed(6);
  const trimmed = s.replace(/0+$/, '');
  return trimmed.endsWith('.') ? trimmed : trimmed;
}

function _esc(s) { return String(s ?? '').replace(/'/g, "''"); }
